heat/core/linalg/solver.py

```diff
@@ -85,100 +86,262 @@ def lanczos(
+                    vr._DNDarray__array = vr._DNDarray__array - a / b * vi_loc
 
-            vi = vr / ht.norm(vr)
+                vi = vr / ht.norm(vr)
 
-        w = ht.matmul(A, vi)
-        alpha = ht.dot(w, vi)
+            w = ht.matmul(A, vi)
+            alpha = ht.dot(w, vi)
 
-        w = w - alpha * vi - beta * V[:, i - 1]
+            w = w - alpha * vi - beta * V[:, i - 1]
 
-        T[i - 1, i] = beta
-        T[i, i - 1] = beta
-        T[i, i] = alpha
-        V[:, i] = vi
+            T[i - 1, i] = beta
+            T[i, i - 1] = beta
+            T[i, i] = alpha
+            V[:, i] = vi
 
     if V.split is not None:
         V.resplit_(axis=None)
 
-    if T_out is not None:
-        T_out = T.copy()
-        if V_out is not None:
-            V_out = V.copy()
-            return V_out, T_out
-        return V, T_out
-    elif V_out is not None:
-        V_out = V.copy()
-        return V_out, T
-
     return V, T
+
+
+def triangular_solve(A: DNDarray, B: DNDarray) -> DNDarray:
+    """
+    Solver for a system of equations :math: `Ax = B`, where A is a triangular matrix
+
+    Parameters
+    ---------
+    A : DNDarray
+        2D symmetric, Triangular Matrix
+    B : DNDarray
+        2D Matrix
+    """
+
+    if not isinstance(A, DNDarray) or not isinstance(B, DNDarray):
+        raise TypeError("A and B need to be of type ht.DNDarray")
+
+    if not A.ndim == 2:
+        raise RuntimeError("A needs to be a 2D matrix")
+
+    (m, m1) = A.shape
+    (m2, n) = B.shape
+    if not (m1 == m):
+        raise TypeError("A must be a square matrix")
+    if not (m == m2):
+        raise TypeError("A and B must have same number of rows")
+
+    type = "Upper"
+    if type == "Upper":
+        for i in range(m):
+            for j in range(i + 1, m):
+                if not A[j, i] == 0:
+                    type = "Lower"
+            if type == "Lower":
+                break
+
+    if type == "Lower":
+        for i in range(m):
+            for j in range(0, i):
+                if not A[j, i] == 0:
+                    type = "Not"
+            if type == "Not":
+                break
+
+    if type == "Not":
+        raise TypeError("The matrix A is not a triangular matrix, please use cg solver")
+
+    for i in range(m):
+        if A[i, i] == 0:
+            raise TypeError("The matrix A is not full rank, hense the system cannot be solved")
+
+    x = ht.zeros(
+        (
+            m,
+            n,
+        )
+    )
+
+    for i in range(n):
+        b = ht.zeros(m)
+        bi = B[:, i]
+        if type == "Upper":
+            for j in range(m - 1, -1, -1):
+                value = ((bi - b)[j]) / A[j, j]
+                b += value * (A[:, j])
+                x[j, i] = value
+        else:
+            for j in range(0, m, +1):
+                value = ((bi - b)[j]) / A[j, j]
+                b += value * (A[:, j])
+                x[j, i] = value
+
```

Comment:
@Gaurav17Joshi this is not going to work, because our matrices are going to be distributed across nodes (or processes, or ranks in MPI terminology). 

That means, if A.shape is (m,m) and we run the solver of 4 nodes:
- the first node (rank 0) will only hold the slice of data A[m/4, m]
- the second (rank 1) will hold A[m/4:m/2, m]
- and so on, and same with B.

We need to research how to break up this solver in:
- process-local operations that can be applied to the local slice of data (these should be `torch` operations)
plus:
- any communication (see the `communication` module) that needs to happen among processes.